### DISCUSSION TITLE
GCM Plugin is not getting loaded for IKEv1 and IKEv2

### DISCUSSION THREAD

**Discussion by manasajamuna** [strongswan/strongswan#826] [Q&A]
Hi, 

I am using Strongswan-4.5.0. --enable-gcm is passed during configure. 

./configure --prefix=/usr --build x86_64-unknown-linux-gnu --host x86_64-wios-linux-gnu --target x86_64-wios-linux-gnu --disable-tools --enable-monolithic --enable-openssl --disable-gmp --enable-dhcp --enable-farp --enable-xcbc --enable-gcm

With this, GCM is built successfully. libstrongswan-gcm.so is created. 

But the output of 'ipsec listalgs' and 'ipsec statusall' do not list 'gcm' as loaded plugin. Also, 'aead' section is empty. 

root # ipsec listalgs

List of registered IKEv1 Algorithms:

  encryption: BLOWFISH_CBC 3DES_CBC AES_CBC CAMELLIA_CBC
  integrity:  HMAC_MD5 HMAC_SHA1 HMAC_SHA2_256 HMAC_SHA2_384 HMAC_SHA2_512
  dh-group:   MODP_1024 MODP_1536 MODP_2048 MODP_3072 MODP_4096 MODP_6144 MODP_8192 ECP_256 ECP_384 ECP_521 MODP_1024_160 MODP_2048_224 MODP_2048_256 ECP_192 ECP_224

List of registered ESP Algorithms:

  encryption: DES_CBC 3DES_CBC NULL AES_CBC AES_GCM_8 AES_GCM_12 AES_GCM_16 AES_GMAC
  integrity:  HMAC_MD5 HMAC_SHA1 HMAC_SHA2_256 NULL

List of registered IKEv2 Algorithms:

  encryption: AES_CBC 3DES_CBC DES_CBC DES_ECB CAMELLIA_CBC RC5_CBC IDEA_CBC CAST_CBC BLOWFISH_CBC NULL
  integrity:  AES_XCBC_96 CAMELLIA_XCBC_96 HMAC_SHA1_96 HMAC_SHA1_128 HMAC_SHA1_160 HMAC_SHA2_256_128 HMAC_SHA2_256_256 HMAC_MD5_96 HMAC_MD5_128 HMAC_SHA2_384_192 HMAC_SHA2_384_384 HMAC_SHA2_512_256
  aead:
  hasher:     HASH_SHA1 HASH_MD5 HASH_SHA224 HASH_SHA256 HASH_SHA384 HASH_SHA512 HASH_MD2 HASH_MD4
  prf:        PRF_AES128_XCBC PRF_CAMELLIA128_XCBC PRF_KEYED_SHA1 PRF_HMAC_SHA2_256 PRF_HMAC_SHA1 PRF_HMAC_MD5 PRF_HMAC_SHA2_384 PRF_HMAC_SHA2_512
  dh-group:   MODP_2048 MODP_2048_224 MODP_2048_256 MODP_1536 ECP_256 ECP_384 ECP_521 ECP_224 ECP_192 MODP_3072 MODP_4096 MODP_6144 MODP_8192 MODP_1024 MODP_1024_160 MODP_768 MODP_CUSTOM

/root # ipsec statusall
loaded plugins: aes des sha1 md5 sha2 hmac openssl xauth pkcs1 pem random kernel-netlink
Status of IKEv2 charon daemon (strongSwan 4.5.0):
  uptime: 22 hours, since Dec 20 07:35:13 2021
  malloc: sbrk 2162688, mmap 0, used 151792, free 2010896
  worker threads: 7 idle of 16, job queue load: 0, scheduled events: 4
  loaded plugins: aes xcbc des sha1 md5 sha2 hmac openssl random x509 pkcs1 pem stroke kernel-netlink socket-raw updown farp dhcp attr resolve

If GCM based algorithms like ‘ENCR_AES_GCM_ICV8’ are passed to Strongswan, it throws below errors - 
%AUTHPRIV-6-INFO: charon: 356701952[CFG] <3> received proposals: IKE:AES_GCM_8_256/PRF_HMAC_SHA1/MODP_1024
%AUTHPRIV-6-INFO: charon: 356701952[CFG] <3> configured proposals: IKE:AES_GCM_8_256/PRF_HMAC_SHA1/MODP_1024 %AUTHPRIV-6-INFO: charon: 356701952[CFG] <3> selected proposal: IKE:AES_GCM_8_256/PRF_HMAC_SHA1/MODP_1024
%AUTHPRIV-6-INFO: charon: 356701952[IKE] <3> ENCRYPTION_ALGORITHM AES_GCM_8 (key size 256) not supported!
%AUTHPRIV-6-INFO: charon: 356701952[IKE] <3> key derivation failed

Same error is observed for ENCR_AES_GCM_ICV12’ and ‘ENCR_AES_GCM_ICV16’ 

Kernel version used is '4.14.142' and 'CONFIG_CRYPTO_GCM' is enabled. 
~ # grep -nr gcm /proc/crypto
1:name         : __generic-gcm-aes-aesni
2:driver       : cryptd(__driver-generic-gcm-aes-aesni)
31:name         : gcm(aes)
32:driver       : gcm_base(ctr(aes-aesni),ghash-generic)
123:name         : gcm(aes)
124:driver       : generic-gcm-aesni
137:name         : __generic-gcm-aes-aesni
138:driver       : __driver-generic-gcm-aes-aesni
151:name         : rfc4106(gcm(aes))
152:driver       : rfc4106-gcm-aesni
165:name         : __gcm-aes-aesni
166:driver       : __driver-gcm-aes-aesni

Can you please let me know why is 'gcm' plugin not getting loaded for IKEv2 and IKEv2 ?

**Comment by Thermi** [comment]
The plugins and `/proc/crypto` have nothing to do with each other.

The plugins are only for usage of the algorithms within strongswan. `/proc/crypto` is only relevant for when you make the kernel process ESP and AH packets.

Further edit: That version (`4.5.0`) is so old, you're entirely on your own there.
